### PR TITLE
[Feature] Support Concat dataset

### PR DESCRIPTION
**Pull request by Canwang-sjtu**
## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by jin-s13 — APPROVED** [review]

File: mmpose/datasets/builder.py [hunk 1]
```diff
@@ -24,6 +26,38 @@
 PIPELINES = Registry('pipeline')
 
 
+def _concat_dataset(cfg, default_args=None):
```
Comment (ly015) [inline mmpose/datasets/builder.py:29]: It seems an OpenMMLab convention that the ConcatDataset is implemented in `dataset/dataset_wrappers.py` and by inheriting the one from PyTorch and adding the subclass into the registry (For example in [mmcls](https://github.com/open-mmlab/mmclassification/blob/32d606e1a110188eb050e2d579120a3277b76957/mmcls/datasets/dataset_wrappers.py#L13) and [mmdet](https://github.com/open-mmlab/mmdetection/blob/ff9bc39913cb3ff5dde79d3933add7dc2561bab7/mmdet/datasets/dataset_wrappers.py#L17)). Can we also follow this practice?

**Comment by ly015** [comment]
Would you be willing to add a unit test for the new feature?

File: mmpose/datasets/builder.py (hunk 1 above)
Comment (jin-s13) [reply mmpose/datasets/builder.py:29]: We do not need to rewrite ConcatDataset. The official one from PyTorch works fine.
Note: mmcls/mmdet also have _concat_dataset in builder.py

File: mmpose/datasets/builder.py [hunk 2]
```diff
@@ -37,9 +71,13 @@ def build_dataset(cfg, default_args=None):
     """
     from .dataset_wrappers import RepeatDataset
 
-    if cfg['type'] == 'RepeatDataset':
+    if isinstance(cfg, (list, tuple)):
+        dataset = ConcatDataset([build_dataset(c, default_args) for c in cfg])
+    elif cfg['type'] == 'RepeatDataset':
         dataset = RepeatDataset(
             build_dataset(cfg['dataset'], default_args), cfg['times'])
+    elif isinstance(cfg.get('ann_file'), (list, tuple)):
```
Comment (ly015) [inline mmpose/datasets/builder.py:79]: What is the difference between L74 and L79? It is to support two different ways to define ConcatDataset in configs?

Comment (jin-s13) [reply mmpose/datasets/builder.py:79]: It follows mmdet

**Review by jin-s13 — APPROVED** [review]

**Comment by chenxinfeng4** [comment]
Error of 'ConcatDataset', saying `'ConcatDataset' object has no attribute 'evaluate'`

I've update the code in 2022-2-22.
```
Traceback (most recent call last):
  File "tools/train.py", line 195, in <module>
    main()
  File "tools/train.py", line 191, in main
    meta=meta)
  File "/DATA/****/mmpose/mmpose/apis/train.py", line 191, in train_model
    runner.run(data_loaders, cfg.workflow, cfg.total_epochs)
  File "/home/****/lib/python3.7/site-packages/mmcv/runner/epoch_based_runner.py", line 127, in run
    epoch_runner(data_loaders[i], **kwargs)
  File "/home/****/.conda/envs/mmpose/lib/python3.7/site-packages/mmcv/runner/epoch_based_runner.py", line 54, in train
    self.call_hook('after_train_epoch')
  File "/home/****/.conda/envs/mmpose/lib/python3.7/site-packages/mmcv/runner/base_runner.py", line 307, in call_hook
    getattr(hook, fn_name)(self)
  File "/home/****/.conda/envs/mmpose/lib/python3.7/site-packages/mmcv/runner/hooks/evaluation.py", line 267, in after_train_epoch
    self._do_evaluate(runner)
  File "/home/****/.conda/envs/mmpose/lib/python3.7/site-packages/mmcv/runner/hooks/evaluation.py", line 505, in _do_evaluate
    key_score = self.evaluate(runner, results)
  File "****/mmpose/mmpose/core/evaluation/eval_hooks.py", line 135, in evaluate
    eval_res = self.dataloader.dataset.evaluate(
AttributeError: 'ConcatDataset' object has no attribute 'evaluate'
```

This is my config file
```
data = dict(
    workers_per_gpu=2,
    train_dataloader=dict(samples_per_gpu=24),
    val_dataloader=dict(samples_per_gpu=1),
    test_dataloader=dict(samples_per_gpu=1),
    train=dict(
        type = 'ConcatDataset',
        datasets = [dict(
            type='BottomUpCocoDataset',
            ann_file='data/rat_keypoints/2021-11-02-bwrat_side6-kp_train.json',
            img_prefix='data/rat_keypoints/2021-11-02-bwrat_side6-kp/',
            data_cfg=data_cfg,
            pipeline=train_pipeline,
            dataset_info=dataset_info),
        dict(
            type='BottomUpCocoDataset',
            ann_file='data/rat_keypoints/2021-11-24-wrat_side6-kp_train.json',
            img_prefix='data/rat_keypoints/2021-11-24-wrat_side6-kp/',
            data_cfg=data_cfg,
            pipeline=train_pipeline,
            dataset_info=dataset_info)
        ]),
    val=dict(
        type = 'ConcatDataset',
        datasets = [dict(type='BottomUpCocoDataset',
            ann_file='data/rat_keypoints/2021-11-02-bwrat_side6-kp_val.json',
            img_prefix='data/rat_keypoints/2021-11-02-bwrat_side6-kp/',
            data_cfg=data_cfg,
            pipeline=val_pipeline,
            dataset_info=dataset_info),
        dict(
            type='BottomUpCocoDataset',
            ann_file='data/rat_keypoints/2021-11-24-wrat_side6-kp_val.json',
            img_prefix='data/rat_keypoints/2021-11-24-wrat_side6-kp/',
            data_cfg=data_cfg,
            pipeline=val_pipeline,
            dataset_info=dataset_info)
        ]),
    test=dict(
        type = 'ConcatDataset',
        datasets = [dict(
            type='BottomUpCocoDataset',
            ann_file='data/rat_keypoints/2021-11-02-bwrat_side6-kp_val.json',
            img_prefix='data/rat_keypoints/2021-11-02-bwrat_side6-kp/',
            data_cfg=data_cfg,
            pipeline=test_pipeline,
            dataset_info=dataset_info),
        dict(
            type='BottomUpCocoDataset',
            ann_file='data/rat_keypoints/2021-11-24-wrat_side6-kp_val.json',
            img_prefix='data/rat_keypoints/2021-11-24-wrat_side6-kp/',
            data_cfg=data_cfg,
            pipeline=val_pipeline,
            dataset_info=dataset_info)
        ])
)
```

**Pull request merged**